File: src/utils/services_excel_builder.py
#!/usr/bin/env python3
"""
Build and read Excel files for services catalog import/export.
"""

from decimal import Decimal, InvalidOperation
from pathlib import Path
from urllib.parse import unquote, urlparse

from openpyxl import Workbook, load_workbook
from openpyxl.styles import Alignment

_BASE_COL_WIDTH = 15.0
_EXPORT_COL1_WIDTH = _BASE_COL_WIDTH * 4
_EXPORT_COL2_WIDTH = _BASE_COL_WIDTH * 2.0
_WRAP_ALIGNMENT = Alignment(wrap_text=True, vertical="top")


def resolve_excel_path(file_url, ensure_extension=False):
    """Resolve local path from plain path or file:// URL."""
    value = (file_url or "").strip()
    if not value:
        return ""

    if value.startswith("file:"):
        parsed = urlparse(value)
        path = unquote(parsed.path)
        if path.startswith("/") and len(path) > 2 and path[2] == ":":
            path = path[1:]
    else:
        path = value

    if ensure_extension and not path.lower().endswith(".xlsx"):
        path += ".xlsx"
    return path


def _cell_text(value):
    if value is None:
        return ""
    return str(value).strip()


def _normalize_price_text(value):
    if value is None or isinstance(value, bool):
        return None

    if isinstance(value, int):
        return str(value)

    if isinstance(value, float):
        if value.is_integer():
            return str(int(value))
        text = f"{value:.10f}".rstrip("0").rstrip(".")
        return text.replace(",", ".")

    text = str(value).strip().replace(",", ".")
    return text or None


def parse_excel_price(value):
    """
    Parse price from Excel cell.

    Empty value means paragraph row.
    Without decimal separator the value is multiplied by 100 (rubles to kopecks).
    With decimal separator rubles and kopecks are parsed from the text.
    """
    text = _normalize_price_text(value)
    if text is None:
        return None

    try:
        if "." not in text:
            return int(Decimal(text) * 100)

        whole, fraction = text.split(".", 1)
        whole = whole or "0"
        fraction = (fraction + "00")[:2]
        return int(whole) * 100 + int(fraction)
    except (InvalidOperation, ValueError):
        return None


def _price_for_db(cents):
    return f"{cents / 100:.2f}"


def _price_for_excel(cents):
    """Convert stored kopecks back to Excel price cell value."""
    value = int(cents or 0)
    if value % 100 == 0:
        return value // 100
    return float(Decimal(value) / 100)


def _group_services_by_paragraph(services):
    grouped = {}
    for service in services:
        paragraph = str(service.get("paragraph") or "").strip()
        grouped.setdefault(paragraph, []).append(service)

    for paragraph in grouped:
        grouped[paragraph].sort(key=lambda item: str(item.get("name") or "").lower())

    paragraphs = sorted(grouped.keys(), key=lambda text: (text == "", text.lower()))
    return [(paragraph, grouped[paragraph]) for paragraph in paragraphs]


def _row_is_empty(cells):
    return all(_cell_text(cell) == "" for cell in cells[:4])


class ServicesExcelBuilder:
    """Read and write services catalog Excel files."""

    def read_services(self, file_path):
        """
        Read services from Excel.

        Row without price is a paragraph applied to following service rows.
        Service row columns: name, note, price, unit.
        """
        path = Path(file_path)
        if not path.is_file():
            raise FileNotFoundError(f"Services import file not found: {path}")

        services = []
        current_paragraph = ""

        workbook = load_workbook(path, read_only=True, data_only=True)
        try:
            sheet = workbook.active
            for row in sheet.iter_rows(values_only=True):
                cells = list(row)
                if len(cells) < 4:
                    cells.extend([None] * (4 - len(cells)))

                if _row_is_empty(cells):
                    continue

                name = _cell_text(cells[0])
                note = _cell_text(cells[1])
                price_cents = parse_excel_price(cells[2])
                unit = _cell_text(cells[3])

                if price_cents is None:
                    paragraph_text = name or note
                    if paragraph_text:
                        current_paragraph = paragraph_text
                    continue

                if not name:
                    continue

                services.append({
                    "name": name,
                    "note": note,
                    "paragraph": current_paragraph,
                    "price": _price_for_db(price_cents),
                    "unit": unit,
                })
        finally:
            workbook.close()

        return services

    def write_export(self, file_path, services):
        """
        Write services catalog to Excel.

        Services are grouped by paragraph; each group starts with a merged header
        row across four columns. Service rows: name, note, price, unit.
        First two columns use wider layout and wrap long text.
        """
        path = Path(file_path)
        path.parent.mkdir(parents=True, exist_ok=True)

        workbook = Workbook()
        sheet = workbook.active
        sheet.column_dimensions["A"].width = _EXPORT_COL1_WIDTH
        sheet.column_dimensions["B"].width = _EXPORT_COL2_WIDTH

        row_num = 1
        try:
            for paragraph, items in _group_services_by_paragraph(services):
                if paragraph:
                    sheet.cell(row=row_num, column=1, value=paragraph)
                    sheet.merge_cells(
                        start_row=row_num,
                        start_column=1,
                        end_row=row_num,
                        end_column=4,
                    )
                    sheet.cell(row=row_num, column=1).alignment = _WRAP_ALIGNMENT
                    row_num += 1

                for service in items:
                    name = str(service.get("name") or "").strip()
                    if not name:
                        continue

                    note = str(service.get("note") or "").strip()
                    price = _price_for_excel(service.get("price", 0))
                    unit = str(service.get("unit") or "").strip()

                    sheet.cell(row=row_num, column=1, value=name)
                    sheet.cell(row=row_num, column=2, value=note or None)
                    sheet.cell(row=row_num, column=3, value=price)
                    sheet.cell(row=row_num, column=4, value=unit or None)
                    sheet.cell(row=row_num, column=1).alignment = _WRAP_ALIGNMENT
                    sheet.cell(row=row_num, column=2).alignment = _WRAP_ALIGNMENT
                    row_num += 1

            workbook.save(path)
        finally:
            workbook.close()
        return str(path.resolve())


def import_services_from_excel(file_url):
    """Read services list from Excel file referenced by path or URL."""
    path = resolve_excel_path(file_url)
    if not path:
        raise ValueError("Services import path is empty")
    return ServicesExcelBuilder().read_services(path)


def export_services_excel(file_url, services=None):
    """Write services catalog to Excel file at the given path or URL."""
    path = resolve_excel_path(file_url, ensure_extension=True)
    if not path:
        raise ValueError("Services export path is empty")

    if services is None:
        from utils.services_db import load_services
        services = load_services()

    return ServicesExcelBuilder().write_export(path, services)

File: cpp/utils/servicesExcelBuilder.cpp
#include "servicesExcelBuilder.h"

#include "servicesDatabase.h"

#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QUrl>
#include <QXmlStreamReader>
#include <QXmlStreamWriter>

#include <algorithm>
#include <cmath>
#include <stdexcept>

#include "minimalZip.h"

namespace workorder
{

namespace
{

const double kBaseColWidth = 15.0;
const double kExportCol1Width = kBaseColWidth * 4.0;
const double kExportCol2Width = kBaseColWidth * 2.0;

QString cellText(const QVariant &value)
{
    if(!value.isValid() || value.isNull())
        return QString();
    return value.toString().trimmed();
}

QString normalizePriceText(const QVariant &value)
{
    if(!value.isValid() || value.typeId() == QMetaType::Bool)
        return QString();

    if(value.typeId() == QMetaType::Int)
        return QString::number(value.toInt());

    if(value.typeId() == QMetaType::Double)
    {
        const double number = value.toDouble();
        if(qFuzzyCompare(number, std::round(number)))
            return QString::number(static_cast<qint64>(std::round(number)));
        QString text = QString::number(number, 'f', 10);
        while(text.endsWith('0'))
            text.chop(1);
        if(text.endsWith('.'))
            text.chop(1);
        return text.replace(',', '.');
    }

    const QString text = value.toString().trimmed().replace(',', '.');
    return text;
}

QString priceForDb(int cents)
{
    return QString::number(cents / 100.0, 'f', 2);
}

QVariant priceForExcel(int cents)
{
    if(cents % 100 == 0)
        return cents / 100;
    return cents / 100.0;
}

using ParagraphGroup = QPair<QString, StringMapList>;

QVector<ParagraphGroup> groupServicesByParagraph(const StringMapList &services)
{
    QMap<QString, StringMapList> grouped;
    for(const QVariant &item : services)
    {
        const StringMap service = item.toMap();
        const QString paragraph = service.value("paragraph").toString().trimmed();
        grouped[paragraph].append(service);
    }

    QStringList paragraphs = grouped.keys();
    std::sort(paragraphs.begin(), paragraphs.end(), [](const QString &a, const QString &b)
    {
        if(a.isEmpty())
            return false;
        if(b.isEmpty())
            return true;
        return a.compare(b, Qt::CaseInsensitive) < 0;
    });

    QVector<ParagraphGroup> result;
    for(const QString &paragraph : paragraphs)
    {
        StringMapList items = grouped.value(paragraph);
        std::sort(items.begin(), items.end(), [](const QVariant &left, const QVariant &right)
        {
            return left.toMap().value("name").toString().compare(
                right.toMap().value("name").toString(), Qt::CaseInsensitive) < 0;
        });
        result.append({paragraph, items});
    }
    return result;
}

QString xmlEscape(const QString &value)
{
    QString text = value;
    text.replace('&', "&amp;");
    text.replace('<', "&lt;");
    text.replace('>', "&gt;");
    text.replace('"', "&quot;");
    return text;
}

QString columnName(int index)
{
    QString result;
    int value = index;
    while(value > 0)
    {
        const int rem = (value - 1) % 26;
        result.prepend(QChar('A' + rem));
        value = (value - 1) / 26;
    }
    return result;
}

QStringList readSharedStrings(const QByteArray &sharedStringsXml)
{
    QStringList strings;
    QXmlStreamReader reader(sharedStringsXml);
    while(!reader.atEnd())
    {
        reader.readNext();
        if(reader.isStartElement() && reader.name() == QLatin1String("t"))
            strings.append(reader.readElementText());
    }
    return strings;
}

StringMapList readSheetRows(const QByteArray &sheetXml, const QStringList &sharedStrings)
{
    StringMapList services;
    QString currentParagraph;
    QXmlStreamReader reader(sheetXml);

    QStringList rowCells(4);
    while(!reader.atEnd())
    {
        reader.readNext();
        if(reader.isStartElement() && reader.name() == QLatin1String("row"))
        {
            rowCells = QStringList({QString(), QString(), QString(), QString()});
        }
        else if(reader.isStartElement() && reader.name() == QLatin1String("c"))
        {
            const QString ref = reader.attributes().value("r").toString();
            const QChar columnChar = ref.isEmpty() ? QChar() : ref.at(0);
            const int column = columnChar.isNull() ? -1 : columnChar.toLatin1() - 'A';
            const QString cellType = reader.attributes().value("t").toString();

            QString value;
            while(!(reader.isEndElement() && reader.name() == QLatin1String("c")))
            {
                reader.readNext();
                if(reader.isStartElement() && reader.name() == QLatin1String("v"))
                    value = reader.readElementText();
                else if(reader.isStartElement() && reader.name() == QLatin1String("t"))
                    value = reader.readElementText();
            }

            if(column >= 0 && column < rowCells.size())
            {
                if(cellType == QLatin1String("s"))
                {
                    bool ok = false;
                    const int index = value.toInt(&ok);
                    rowCells[column] = ok ? sharedStrings.value(index) : value;
                }
                else
                {
                    rowCells[column] = value;
                }
            }
        }
        else if(reader.isEndElement() && reader.name() == QLatin1String("row"))
        {
            if(rowCells.at(0).isEmpty() && rowCells.at(1).isEmpty()
                && rowCells.at(2).isEmpty() && rowCells.at(3).isEmpty())
            {
                continue;
            }

            const int priceCents = ServicesExcelBuilder::parseExcelPrice(rowCells.at(2));
            if(priceCents < 0)
            {
                const QString paragraphText = !rowCells.at(0).isEmpty() ? rowCells.at(0) : rowCells.at(1);
                if(!paragraphText.isEmpty())
                    currentParagraph = paragraphText;
                continue;
            }

            if(rowCells.at(0).isEmpty())
                continue;

            StringMap service;
            service.insert("name", rowCells.at(0));
            service.insert("note", rowCells.at(1));
            service.insert("paragraph", currentParagraph);
            service.insert("price", priceForDb(priceCents));
            service.insert("unit", rowCells.at(3));
            services.append(service);
        }
    }

    return services;
}

QString buildSheetXml(const StringMapList &services)
{
    QString xml;
    QXmlStreamWriter writer(&xml);
    writer.setAutoFormatting(true);
    writer.writeStartDocument();
    writer.writeStartElement("worksheet");
    writer.writeDefaultNamespace("http://schemas.openxmlformats.org/spreadsheetml/2006/main");
    writer.writeStartElement("cols");
    writer.writeEmptyElement("col");
    writer.writeAttribute("min", "1");
    writer.writeAttribute("max", "1");
    writer.writeAttribute("width", QString::number(kExportCol1Width));
    writer.writeAttribute("customWidth", "1");
    writer.writeEmptyElement("col");
    writer.writeAttribute("min", "2");
    writer.writeAttribute("max", "2");
    writer.writeAttribute("width", QString::number(kExportCol2Width));
    writer.writeAttribute("customWidth", "1");
    writer.writeEndElement();

    writer.writeStartElement("sheetData");
    int rowNum = 1;
    for(const ParagraphGroup &group : groupServicesByParagraph(services))
    {
        if(!group.first.isEmpty())
        {
            writer.writeStartElement("row");
            writer.writeAttribute("r", QString::number(rowNum));
            writer.writeStartElement("c");
            writer.writeAttribute("r", QString("%1%2").arg(columnName(1), QString::number(rowNum)));
            writer.writeAttribute("t", "inlineStr");
            writer.writeStartElement("is");
            writer.writeTextElement("t", group.first);
            writer.writeEndElement();
            writer.writeEndElement();
            writer.writeEndElement();
            ++rowNum;
        }

        for(const QVariant &item : group.second)
        {
            const StringMap service = item.toMap();
            const QString name = service.value("name").toString().trimmed();
            if(name.isEmpty())
                continue;

            writer.writeStartElement("row");
            writer.writeAttribute("r", QString::number(rowNum));

            const QStringList values = {
                name,
                service.value("note").toString().trimmed(),
                priceForExcel(service.value("price").toInt()).toString(),
                service.value("unit").toString().trimmed(),
            };

            for(int col = 0; col < values.size(); ++col)
            {
                writer.writeStartElement("c");
                writer.writeAttribute("r", QString("%1%2").arg(columnName(col + 1), QString::number(rowNum)));
                if(col == 2 && !values.at(col).isEmpty())
                {
                    writer.writeStartElement("v");
                    writer.writeCharacters(values.at(col));
                    writer.writeEndElement();
                }
                else if(!values.at(col).isEmpty())
                {
                    writer.writeAttribute("t", "inlineStr");
                    writer.writeStartElement("is");
                    writer.writeTextElement("t", values.at(col));
                    writer.writeEndElement();
                }
                writer.writeEndElement();
            }

            writer.writeEndElement();
            ++rowNum;
        }
    }
    writer.writeEndElement();
    writer.writeEndElement();
    writer.writeEndDocument();
    return xml;
}

QByteArray buildWorkbookXml()
{
    return QByteArray(
        "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>"
        "<workbook xmlns=\"http://schemas.openxmlformats.org/spreadsheetml/2006/main\" "
        "xmlns:r=\"http://schemas.openxmlformats.org/officeDocument/2006/relationships\">"
        "<sheets><sheet name=\"Services\" sheetId=\"1\" r:id=\"rId1\"/></sheets>"
        "</workbook>");
}

} // namespace

QString ServicesExcelBuilder::resolveExcelPath(const QString &fileUrl, bool ensureExtension)
{
    QString value = fileUrl.trimmed();
    if(value.isEmpty())
        return QString();

    QString path;
    if(value.startsWith("file:", Qt::CaseInsensitive))
        path = QUrl(value).toLocalFile();
    else
        path = value;

    if(ensureExtension && !path.endsWith(".xlsx", Qt::CaseInsensitive))
        path += ".xlsx";

    return path;
}

int ServicesExcelBuilder::parseExcelPrice(const QVariant &value)
{
    const QString text = normalizePriceText(value);
    if(text.isEmpty())
        return -1;

    if(!text.contains('.'))
    {
        bool ok = false;
        const qlonglong whole = text.toLongLong(&ok);
        return ok ? static_cast<int>(whole * 100) : -1;
    }

    const QStringList parts = text.split('.');
    if(parts.size() != 2)
        return -1;

    bool wholeOk = false;
    bool fractionOk = false;
    const qlonglong whole = parts.at(0).isEmpty() ? 0 : parts.at(0).toLongLong(&wholeOk);
    QString fraction = parts.at(1);
    fraction = (fraction + "00").left(2);
    const int fractionValue = fraction.toInt(&fractionOk);
    if(!wholeOk || !fractionOk)
        return -1;

    return static_cast<int>(whole * 100 + fractionValue);
}

StringMapList ServicesExcelBuilder::readServices(const QString &filePath)
{
    QFile file(filePath);
    if(!file.exists())
        throw std::runtime_error(QString("Services import file not found: %1").arg(filePath).toStdString());

    if(!file.open(QIODevice::ReadOnly))
        throw std::runtime_error(QString("Failed to open services import file: %1").arg(filePath).toStdString());

    MinimalZipReader zip(file.readAll());
    if(!zip.isValid())
        throw std::runtime_error("Invalid XLSX archive");

    const QStringList sharedStrings = readSharedStrings(zip.fileContent("xl/sharedStrings.xml"));
    const QByteArray sheetXml = zip.fileContent("xl/worksheets/sheet1.xml");
    if(sheetXml.isEmpty())
        throw std::runtime_error("Worksheet sheet1.xml not found in XLSX");

    return readSheetRows(sheetXml, sharedStrings);
}

QString ServicesExcelBuilder::writeExport(const QString &filePath, const StringMapList &services)
{
    QDir().mkpath(QFileInfo(filePath).absolutePath());

    const QByteArray contentTypes =
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
        "<Types xmlns=\"http://schemas.openxmlformats.org/package/2006/content-types\">"
        "<Default Extension=\"rels\" ContentType=\"application/vnd.openxmlformats-package.relationships+xml\"/>"
        "<Default Extension=\"xml\" ContentType=\"application/xml\"/>"
        "<Override PartName=\"/xl/workbook.xml\" "
        "ContentType=\"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml\"/>"
        "<Override PartName=\"/xl/worksheets/sheet1.xml\" "
        "ContentType=\"application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml\"/>"
        "</Types>";

    const QByteArray rootRels =
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
        "<Relationships xmlns=\"http://schemas.openxmlformats.org/package/2006/relationships\">"
        "<Relationship Id=\"rId1\" Type=\"http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument\" "
        "Target=\"xl/workbook.xml\"/>"
        "</Relationships>";

    const QByteArray workbookRels =
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
        "<Relationships xmlns=\"http://schemas.openxmlformats.org/package/2006/relationships\">"
        "<Relationship Id=\"rId1\" Type=\"http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet\" "
        "Target=\"worksheets/sheet1.xml\"/>"
        "</Relationships>";

    MinimalZipWriter zip;
    zip.addFile("[Content_Types].xml", contentTypes);
    zip.addFile("_rels/.rels", rootRels);
    zip.addFile("xl/workbook.xml", buildWorkbookXml());
    zip.addFile("xl/_rels/workbook.xml.rels", workbookRels);
    zip.addFile("xl/worksheets/sheet1.xml", buildSheetXml(services).toUtf8());

    if(!zip.writeToFile(filePath))
        throw std::runtime_error(QString("Failed to write services export file: %1").arg(filePath).toStdString());

    return QFileInfo(filePath).absoluteFilePath();
}

StringMapList importServicesFromExcel(const QString &fileUrl)
{
    const QString path = ServicesExcelBuilder::resolveExcelPath(fileUrl);
    if(path.isEmpty())
        throw std::runtime_error("Services import path is empty");
    return ServicesExcelBuilder().readServices(path);
}

QString exportServicesExcel(const QString &fileUrl, const StringMapList &services)
{
    const QString path = ServicesExcelBuilder::resolveExcelPath(fileUrl, true);
    if(path.isEmpty())
        throw std::runtime_error("Services export path is empty");

    StringMapList data = services;
    if(data.isEmpty())
        data = ServicesDatabase::loadServices();

    return ServicesExcelBuilder().writeExport(path, data);
}

} // namespace workorder
